modules/hibernate-core/src/main/java/org/apache/ignite/cache/hibernate/HibernateAccessStrategyFactory.java

```diff
@@ -188,7 +208,26 @@ private LazyCacheSupplier(String cacheName, String regionName) {
             IgniteInternalCache<Object, Object> cache = reference.get();
 
             if (cache == null) {
-                cache = ((IgniteKernal)ignite).getCache(cacheName);
+                IgniteKernal ignite = (IgniteKernal)HibernateAccessStrategyFactory.this.ignite;
+
+                IgniteCacheProxy<Object, Object> jcache = (IgniteCacheProxy<Object, Object>)ignite.cache(cacheName);
+
+                if (jcache == null) {
+                    try {
+                        CacheConfiguration<?, ?> template = ignite.context().cache().getConfigFromTemplate(cacheName);
+
+                        if (template != null) {
+                            ignite.createCache(template);
+
+                            cache = ignite.getCache(cacheName);
+                        }
+                    }
+                    catch (IgniteCheckedException e) {
+                        throw new IllegalArgumentException("Failed to get cache: " + e.getMessage(), e);
+                    }
+                }
+                else
+                    cache = jcache.internalProxy();
```

Comment:
why do you use `internalProxy` here? What does it do?

Reply:
We need IgniteInternalCache as HibernateCacheProxy relies on IgniteInternalCache (because HibernateCacheProxy is basically a wrapper for InternalCache -- `public class HibernateCacheProxy implements IgniteInternalCache<Object, Object>`)